File: frontend/src/api/notifications.ts
import { api } from "./client";

/**
 * What produced a notification. Drives the icon and the grouping, never the
 * delivery.
 *
 * **All nine the backend defines.** This listed six until spec 065 — `boss_kill`,
 * `announcement` and `dispatch` were missing, so anything built from this union
 * would have silently omitted announcements. A backend test asserts the two
 * lists match rather than trusting them to.
 */
export const NOTIFICATION_KINDS = [
  "achievement",
  "class_unlocked",
  "zone_unlocked",
  "level_up",
  "ability_milestone",
  "boss_kill",
  "announcement",
  "dispatch",
  "system",
] as const;

export type NotificationKind = (typeof NOTIFICATION_KINDS)[number];

export interface AppNotification {
  id: string;
  kind: NotificationKind;
  title: string;
  body: string;
  link: string | null;
  read: boolean;
  created_at: string;
}

export interface NotificationFeed {
  unread: number;
  items: AppNotification[];
}

export const getNotifications = () => api.get<NotificationFeed>("/notifications");

export const markAllRead = () =>
  api.post<{ message: string }>("/notifications/read");

export const markRead = (id: string) =>
  api.post<{ message: string }>(`/notifications/${id}/read`);

/** Clear one row from the inbox. A soft dismiss server-side (spec 065 §4). */
export const dismissOne = (id: string) =>
  api.post<{ message: string }>(`/notifications/${id}/dismiss`);

/** Clear everything, or everything of the given kinds. Scoped so one tab's
 *  clear-all cannot take the other tab's rows with it. */
export const dismissMany = (kinds: NotificationKind[] = []) =>
  api.post<{ message: string }>("/notifications/dismiss", { kinds });

/** Null until earned — redacted server-side, so there is nothing here to
 *  un-blur in devtools (spec 028). */
export interface AchievementRow {
  id: string;
  name: string | null;
  description: string | null;
  earned: boolean;
  rarity: number | null;
}

export interface AchievementsResponse {
  earned: number;
  total: number;
  items: AchievementRow[];
  rarest: AchievementRow[];
}

export const getAchievements = () =>
  api.get<AchievementsResponse>("/character/achievements");

File: backend/app/schemas/notifications.py
"""Notification and achievement responses (spec 028)."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.models.notification import NotificationKind


class DismissRequest(BaseModel):
    """What to clear (spec 065 §4).

    Empty means everything. A kind list scopes it, so "clear all" on one inbox
    tab cannot take the other tab's rows with it.
    """

    kinds: list[NotificationKind] = []


class NotificationResponse(BaseModel):
    id: UUID
    kind: str
    title: str
    body: str
    link: str | None
    read: bool
    created_at: datetime


class NotificationFeed(BaseModel):
    unread: int
    items: list[NotificationResponse]


class AchievementResponse(BaseModel):
    id: UUID
    #: Null until earned. Redacted server-side rather than blurred in CSS, so
    #: the mystery survives devtools.
    name: str | None
    description: str | None
    earned: bool
    #: Share of players who have solved something and hold this, 0..1.
    rarity: float | None


class AchievementsResponse(BaseModel):
    earned: int
    total: int
    items: list[AchievementResponse]
    #: The player's rarest earned achievements — their bragging rights.
    rarest: list[AchievementResponse]


class StarResponse(BaseModel):
    """A boss kill (spec 031). Derived from the solve, never stored."""

    challenge_id: UUID
    #: The stable key (spec 059 §3) — what a client keys a star on.
    slug: str
    challenge_title: str
    zone_name: str
    tier: str
    #: 1 (Neighborhood) to 6 (Floor), so a client can order without the names.
    level: int
